File: src/org/jgll/grammar/CharacterClass.java
package org.jgll.grammar;

import java.util.BitSet;
import java.util.Collections;
import java.util.List;

/**
 * Character class represents a set of {@link Range} instances.
 * For example, [A-Za-z0-9] represents a character which is
 * either [A-Z], [a-z] or [0-9].
 * 
 * @author Ali Afroozeh
 *
 */
public class CharacterClass implements Terminal {
	
	private static final long serialVersionUID = 1L;

	private final List<Range> ranges;
	
	private BitSet testSet;
	
	private int minValue = Integer.MAX_VALUE;
	
	private int end;

	public CharacterClass(List<Range> ranges) {
		if(ranges == null || ranges.size() == 0) {
			throw new IllegalArgumentException("Ranges cannot be null or empty.");
		}
		
		testSet = new BitSet();
		
		for(Range range : ranges) {
			if(minValue > range.getMinimumValue()) {
				minValue = range.getMinimumValue();
			}
			if(end < range.getMaximumValue()) {
				end = range.getMaximumValue();
			}
			testSet.or(range.getTestSet());
		}
		
		this.ranges = Collections.unmodifiableList(ranges);
	}
	
	public List<Range> getRanges() {
		return ranges;
	}
	
	@Override
	public boolean match(int i) {
		if(i < minValue || i > end) {
			return false;
		}
		return testSet.get(i);
	}
	
	@Override
	public String toString() {
		return getName();
	}

	@Override
	public String getMatchCode() {
		StringBuilder sb = new StringBuilder();
		for(Range range : ranges) {
			sb.append(range.getMatchCode()).append(" || ");
		}
		return sb.toString();
	}
	
	@Override
	public int hashCode() {
		return ranges.hashCode();
	}

	@Override
	public boolean equals(Object obj) {
		if (this == obj) {
			return true;
		}

		if (!(obj instanceof CharacterClass)) {
			return false;
		}
		
		CharacterClass other = (CharacterClass) obj;

		return testSet.equals(other.testSet);
	}

	@Override
	public boolean isTerminal() {
		return true;
	}

	@Override
	public boolean isNonterminal() {
		return false;
	}

	@Override
	public String getName() {
		StringBuilder sb = new StringBuilder();
		
		sb.append("[");
		for(Range range : ranges) {
			sb.append(getChar(range.getStart())).append("-").append(getChar(range.getEnd()));
		}
		sb.append("]");
		return sb.toString();
	}
	
	private String getChar(int val) {
		char c = (char) val;
		if(c == '-' || c == ' ') {
			return "\\" + c;
		}
		if(c == '\r') {
			return "\\r";
		}
		if(c == '\n') {
			return "\\n";
		}
		return c + "";
	}
	
	@Override
	public BitSet getTestSet() {
		return testSet;
	}

	@Override
	public int getMinimumValue() {
		return minValue;
	}

	@Override
	public int getMaximumValue() {
		return end;
	}

}

File: src-scala/org/jgll/grammar/CharacterClass.scala
package org.jgll.grammar

import java.util.BitSet
import org.jgll.grammar.condition.Condition
import scala.reflect.BeanProperty
import collection.mutable._
import scala.virtualization.lms.common.Base

trait CharacterClassTrait {
  self: TerminalTrait
   with RangeTrait
   with Base =>
  @SerialVersionUID(1L)
  class CharacterClass(@BeanProperty val ranges: ListBuffer[Range]) extends AbstractSymbol with Terminal {

    private var testSet: BitSet = new BitSet()

    if (ranges == null || ranges.size == 0) {
      throw new IllegalArgumentException("Ranges cannot be null or empty.")
    }

    for (range <- ranges) {
      testSet.or(range.asBitSet())
    }

    override def `match`(i: Rep[Int]): Boolean = testSet.get(i)

    override def toString(): String = getName

    override def getMatchCode(): String = {
      val sb = new StringBuilder()
      for (range <- ranges) {
        sb.append(range.getMatchCode).append(" || ")
      }
      sb.toString
    }

    override def hashCode(): Int = ranges.hashCode

    override def equals(obj: Any): Boolean = {
      if (!(obj.isInstanceOf[CharacterClass])) {
        return false
      }
      val other = obj.asInstanceOf[CharacterClass]
      testSet == other.testSet
    }

    override def getName(): String = {
      val sb = new StringBuilder()
      sb.append("[")
      for (range <- ranges) {
        sb.append(getChar(range.getStart)).append("-").append(getChar(range.getEnd))
      }
      sb.append("]")
      sb.toString
    }

    private def getChar(`val`: Int): String = {
      val c = `val`.toChar
      if (c == '-' || c == ' ') {
        return "\\" + c
      }
      if (c == '\r') {
        return "\\r"
      }
      if (c == '\n') {
        return "\\n"
      }
      c + ""
    }

    override def asBitSet(): BitSet = testSet

    override def addConditions(conditions: ListBuffer[Condition]): Terminal = {
      val characterClass = new CharacterClass(this.ranges)
      characterClass.conditions.++=(this.conditions)
      characterClass.conditions.++=(conditions)
      characterClass
    }
  }
}